jax/_src/random.py

```diff
@@ -1724,3 +1724,85 @@ def ball(
   g = generalized_normal(k1, p, (*shape, d), dtype)
   e = exponential(k2, shape, dtype)
   return g / (((jnp.abs(g) ** p).sum(-1) + e) ** (1 / p))[..., None]
+
+def vonmises(
+  key: KeyArray,
+  x: float,
+  kappa: float,
+  shape: Shape = (),
+  dtype: DTypeLikeFloat = dtypes.float_
+) -> Array:
+  """Draw samples from a von Mises distribution.
+
+  Reference:
+    Best, D. J., & Fisher, N. I. (1979). Efficient Simulation of the
+    von Mises Distribution. Applied Statistics, 28(2), 152.
+    https://doi.org/10.2307/2346732
+
+  Args:
+    key: a PRNG key used as the random key.
+    x: mean of the distribution.
+    kappa: concentration of the distribution.
+    shape: shape of the output array.
+    dtype: dtype of the output array.
+
+  Returns:
+    theta: A random array of specified shape and dtype."""
+  # Run Checks
+  key, _ = _check_prng_key(key)
+  if not dtypes.issubdtype(dtype, np.floating):
+    raise ValueError(f"dtype argument to `t` must be a float"
+                     f"dtype, got {dtype}.")
+  dtype = dtypes.canonicalize_dtype(dtype)
+  shape = core.canonicalize_shape(shape)
+
+  # Return Distribution
+  if kappa <= 1e-6:
+    return ((jnp.pi * 2) * uniform(key, shape, dtype)).astype(dtype)
+  else:
+    return _vonmises(key, x, kappa, shape, dtype, 100)
+
+@partial(jit, static_argnums = (3, 4, 5), inline = True)
+def _vonmises(key, x, kappa, shape, dtype, max_iters):
+
+  def _body_fn(pars):
+    iteration, key, theta_out, accepted = pars
+    key, subkey1, subkey2, subkey3 = _split(key, 4)
+
+    u1 = uniform(subkey1, shape, dtype)
+    u2 = uniform(subkey2, shape, dtype)
+    u3 = uniform(subkey3, shape, dtype)
+    z = lax.cos(jnp.pi * u1)
+    f = (one + r * z) / (r + z)
+    c = kappa * (r - f)
+
+    accepted = lax.ge((c * (two - c) - u2), zero) | lax.ge((lax.log(c / u2) + one) - c, zero)
+    theta = jnp.where(
+      u3 > half,
+      x + jnp.arccos(f) % twopi,
+      x - jnp.arccos(f) % twopi
+    )
+    theta_out = lax.select(accepted, theta, theta_out)
+
+    return iteration + 1, key, theta_out, accepted
+
+  def _cond_fn(pars):
+    iteration, _, _, accepted = pars
+    return (~accepted).any() & (iteration < max_iters)
+
+  x = lax.convert_element_type(x, dtype)
+  kappa = lax.convert_element_type(kappa, dtype)
+  zero = _lax_const(x, 0)
+  half = _lax_const(x, 0.5)
+  one = _lax_const(x, 1)
+  two = _lax_const(x, 2)
+  four = _lax_const(x, 4)
+  twopi = _lax_const(x, (2 * jnp.pi))
```

Comment:
For the most part these constants are unnecessary, and you can just use the literal values directly. The one exception is when you write `lax.ge(expression, zero)` because lax functions do not implicitly promote inputs, but you could use `expression >= 0` instead to avoid the need for strict type matching of inputs.